server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java

```diff
@@ -242,59 +262,93 @@ public Object persist(final Committer committer) throws InterruptedException
   }
 
   /**
-   * Publish all data indexed through this driver so far, and waits for it to be handed off. Blocks until complete.
-   * Retries forever on transient failures, but may exit early on permanent failures.
-   *
-   * Should be called after all data has been added and persisted through {@link #add(InputRow, String, Supplier)} and
-   * {@link #persist(Committer)}.
+   * Register the segments in the given {@link SegmentsAndMetadata} to be handed off and execute a background task which
+   * waits until the hand off completes.
    *
-   * @param publisher publisher to use for this set of segments
-   * @param committer committer representing all data that has been added so far
-   *
-   * @return segments and metadata published if successful, or null if segments could not be handed off due to
-   * transaction failure with commit metadata.
+   * @param segmentsAndMetadata
+   * @return
+   * @throws InterruptedException
    */
-  public SegmentsAndMetadata finish(
-      final TransactionalSegmentPublisher publisher,
-      final Committer committer
-  ) throws InterruptedException
+  public ListenableFuture<SegmentsAndMetadata> registerHandoff(SegmentsAndMetadata segmentsAndMetadata)
```

Comment:
> Also lets see what @gianm has to say

Could you elaborate a bit more on how you imagine the Kafka task using the proposed `registerHandoff` and `waitForHandoff` methods? I think I don't really understand it enough to comment right now.

> Hmm, is it required to call get() on ListenableFutures which are the results of registerHandoff() anyway because that's the only way to detect failures during Appenderator.drop()?

The driver currently suppresses those failures anyway  -- it turns them into alerts and log messages. Although maybe it makes sense to bubble them up and fail the task, on the grounds that will cause the task to get restarted by the supervisor on a maybe-healthier node. I'm not really sure what's best.